File: apps/web/src/lib/run_heatmap.test.ts
import { test } from 'node:test';
import assert from 'node:assert/strict';
import {
	buildHeatCells,
	heatBounds,
	toHeatGeoJSON,
	DEFAULT_GRID_DEG,
	MAX_CELL_WEIGHT,
} from './run_heatmap';

test('collapses many points within one grid cell into a single weighted cell', () => {
	// 10 points within ~10 m of each other → one cell, weight 10.
	const track = Array.from({ length: 10 }, (_, i) => ({
		lat: 51.5 + i * 0.00001,
		lng: -0.12 + i * 0.00001,
	}));
	const cells = buildHeatCells([track]);
	assert.equal(cells.length, 1);
	assert.equal(cells[0].weight, 10);
});

test('repeated routes accumulate weight in shared cells', () => {
	const track = [
		{ lat: 40.0, lng: -75.0 },
		{ lat: 40.001, lng: -75.0 },
	];
	const single = buildHeatCells([track]);
	const tripled = buildHeatCells([track, track, track]);
	// Same cell footprint, 3x the weight.
	assert.equal(single.length, tripled.length);
	const sumSingle = single.reduce((s, c) => s + c.weight, 0);
	const sumTripled = tripled.reduce((s, c) => s + c.weight, 0);
	assert.equal(sumTripled, sumSingle * 3);
});

test('distinct locations produce distinct cells', () => {
	const cells = buildHeatCells([
		[{ lat: 51.5, lng: -0.12 }],
		[{ lat: 48.85, lng: 2.35 }],
	]);
	assert.equal(cells.length, 2);
});

test('clamps a single cells weight at MAX_CELL_WEIGHT', () => {
	const track = Array.from({ length: MAX_CELL_WEIGHT + 200 }, () => ({
		lat: 34.05,
		lng: -118.24,
	}));
	const cells = buildHeatCells([track]);
	assert.equal(cells.length, 1);
	assert.equal(cells[0].weight, MAX_CELL_WEIGHT);
});

test('drops invalid / out-of-range points', () => {
	const cells = buildHeatCells([
		[
			{ lat: NaN, lng: 0 },
			{ lat: 0, lng: Infinity },
			{ lat: 200, lng: 0 },
			{ lat: 0, lng: -400 },
			{ lat: 45, lng: 9 }, // the only valid one
		] as never,
	]);
	assert.equal(cells.length, 1);
	assert.equal(cells[0].weight, 1);
});

test('empty / all-invalid input yields no cells and null bounds', () => {
	assert.deepEqual(buildHeatCells([]), []);
	assert.deepEqual(buildHeatCells([[]]), []);
	assert.equal(heatBounds([]), null);
});

test('heatBounds spans the cell extent as [[w,s],[e,n]]', () => {
	const cells = buildHeatCells([
		[
			{ lat: 10, lng: 20 },
			{ lat: 30, lng: 40 },
		],
	]);
	const b = heatBounds(cells)!;
	assert.ok(b);
	const [[w, s], [e, n]] = b;
	// Cells snap to grid centres, so each edge can sit within one grid
	// step of the raw extent.
	const tol = DEFAULT_GRID_DEG;
	assert.ok(Math.abs(w - 20) <= tol && Math.abs(s - 10) <= tol);
	assert.ok(Math.abs(e - 40) <= tol && Math.abs(n - 30) <= tol);
	assert.ok(w <= e && s <= n);
});

test('invalid gridDeg falls back to the default', () => {
	const track = [{ lat: 1, lng: 1 }];
	assert.deepEqual(buildHeatCells([track], 0), buildHeatCells([track], DEFAULT_GRID_DEG));
	assert.deepEqual(buildHeatCells([track], -5), buildHeatCells([track], DEFAULT_GRID_DEG));
});

test('toHeatGeoJSON emits one weighted point feature per cell', () => {
	const cells = buildHeatCells([
		[
			{ lat: 1, lng: 1 },
			{ lat: 1, lng: 1 },
			{ lat: 2, lng: 2 },
		],
	]);
	const fc = toHeatGeoJSON(cells);
	assert.equal(fc.type, 'FeatureCollection');
	assert.equal(fc.features.length, cells.length);
	for (const f of fc.features) {
		assert.equal(f.geometry.type, 'Point');
		assert.equal(f.geometry.coordinates.length, 2);
		assert.ok(typeof f.properties.weight === 'number' && f.properties.weight >= 1);
	}
});

File: apps/mobile_android/test/run_heatmap_test.dart
// Mirror of apps/web/src/lib/routes/run_heatmap.test.ts for the ported
// pure aggregation (buildHeatCells / heatBounds). The GeoJSON emitters
// on the web side are MapLibre-specific and have no Dart twin, so the
// two toGeoJSON cases are not mirrored here.

import 'package:flutter_test/flutter_test.dart';

import '../lib/run_heatmap.dart';

void main() {
  test('collapses many points within one grid cell into a single weighted cell',
      () {
    final track = [
      for (var i = 0; i < 10; i++)
        HeatLatLng(51.5 + i * 0.00001, -0.12 + i * 0.00001),
    ];
    final cells = buildHeatCells([track]);
    expect(cells.length, 1);
    expect(cells.first.weight, 10);
  });

  test('repeated routes accumulate weight in shared cells', () {
    final track = [
      const HeatLatLng(40.0, -75.0),
      const HeatLatLng(40.001, -75.0),
    ];
    final single = buildHeatCells([track]);
    final tripled = buildHeatCells([track, track, track]);
    expect(single.length, tripled.length);
    final sumSingle = single.fold<int>(0, (s, c) => s + c.weight);
    final sumTripled = tripled.fold<int>(0, (s, c) => s + c.weight);
    expect(sumTripled, sumSingle * 3);
  });

  test('distinct locations produce distinct cells', () {
    final cells = buildHeatCells([
      [const HeatLatLng(51.5, -0.12)],
      [const HeatLatLng(48.85, 2.35)],
    ]);
    expect(cells.length, 2);
  });

  test('clamps a single cell weight at kMaxCellWeight', () {
    final track = [
      for (var i = 0; i < kMaxCellWeight + 200; i++)
        const HeatLatLng(34.05, -118.24),
    ];
    final cells = buildHeatCells([track]);
    expect(cells.length, 1);
    expect(cells.first.weight, kMaxCellWeight);
  });

  test('drops invalid / out-of-range points', () {
    final cells = buildHeatCells([
      [
        const HeatLatLng(double.nan, 0),
        const HeatLatLng(0, double.infinity),
        const HeatLatLng(200, 0),
        const HeatLatLng(0, -400),
        const HeatLatLng(45, 9),
      ],
    ]);
    expect(cells.length, 1);
    expect(cells.first.weight, 1);
  });

  test(
      'quantises a negative coordinate on a grid half-boundary toward +Infinity (cross-platform parity)',
      () {
    // lng/lat = -0.5 * gridDeg sits exactly on a grid half-boundary. The
    // quantiser must round half toward +Infinity (gx = 0 → cell centre 0),
    // matching JS Math.round on the web twin — NOT Dart's raw .round() (which
    // rounds half away from zero to gx = -1). A drift here shifts the cell by
    // one grid step (~33 m) for any Americas / west-of-Greenwich track and the
    // web + mobile heatmaps disagree.
    const g = kDefaultGridDeg;
    final cells = buildHeatCells([
      [const HeatLatLng(-0.5 * g, -0.5 * g)],
    ]);
    expect(cells.length, 1);
    expect(cells.first.lat, 0);
    expect(cells.first.lng, 0);
  });

  test('empty / all-invalid input yields no cells and null bounds', () {
    expect(buildHeatCells([]), isEmpty);
    expect(buildHeatCells([[]]), isEmpty);
    expect(heatBounds([]), isNull);
  });

  test('heatBounds spans the cell extent as [[w,s],[e,n]]', () {
    final cells = buildHeatCells([
      [const HeatLatLng(10, 20), const HeatLatLng(30, 40)],
    ]);
    final b = heatBounds(cells)!;
    final w = b[0][0], s = b[0][1], e = b[1][0], n = b[1][1];
    const tol = kDefaultGridDeg;
    expect((w - 20).abs() <= tol && (s - 10).abs() <= tol, isTrue);
    expect((e - 40).abs() <= tol && (n - 30).abs() <= tol, isTrue);
    expect(w <= e && s <= n, isTrue);
  });

  test('invalid gridDeg falls back to the default', () {
    final track = [const HeatLatLng(1, 1)];
    expect(
      buildHeatCells([track], gridDeg: 0).map((c) => c.weight),
      buildHeatCells([track], gridDeg: kDefaultGridDeg).map((c) => c.weight),
    );
    expect(
      buildHeatCells([track], gridDeg: -5).first.lat,
      buildHeatCells([track], gridDeg: kDefaultGridDeg).first.lat,
    );
  });
}
